HARK/core.py

```diff
@@ -18,11 +19,201 @@
+            setattr(self, key, value)
+
+    def __infer_dims__(self, key, value):
+        """
+        Infers the age-varying dimensions of a parameter.
+
+        If the parameter is a scalar, numpy array, or None, it is assumed to be
+        invariant over time. If the parameter is a list or tuple, it is assumed
+        to be varying over time. If the parameter is a list or tuple of length
+        greater than 1, the length of the list or tuple must match the
+        `_term_age` attribute of the Parameters object.
+
+        Parameters
+        ----------
+        key : str
+            name of parameter
+        value : Any
+            value of parameter
+
+        """
+        if isinstance(value, (int, float, np.ndarray, type(None))):
+            self.__add_to_time_inv(key)
+            return value
+        if isinstance(value, (list, tuple)):
+            if len(value) == 1:
+                self.__add_to_time_inv(key)
+                return value[0]
+            if self._term_age is None or self._term_age == 1:
+                self._term_age = len(value)
+            if len(value) == self._term_age:
+                self.__add_to_time_vary(key)
+                return value
+            raise ValueError(f"Parameter {key} must be of length 1 or {self._term_age}")
+        raise ValueError(f"Parameter {key} has type {type(value)}")
+
+    def __add_to_time_inv(self, key):
+        """
+        Adds parameter name to invariant list and removes from varying list.
+
+        Parameters
+        ----------
+        key : str
+            parameter name
+        """
+        if key in self._age_var:
+            self._age_var.remove(key)
+        if key not in self._age_inv:
+            self._age_inv.append(key)
+
+    def __add_to_time_vary(self, key):
+        """
+        Adds parameter name to varying list and removes from invariant list.
+
+        Parameters
+        ----------
+        key : str
+            parameter name
+        """
+        if key in self._age_inv:
+            self._age_inv.remove(key)
+        if key not in self._age_var:
+            self._age_var.append(key)
+
+    def __getitem__(self, age_or_key):
+        """
+        If age_or_key is an integer, returns a Parameters object with the parameters
+        that apply to that age. This includes all invariant parameters and the
+        `age_or_key`th element of all age-varying parameters. If age_or_key is a string,
+        it returns the value of the parameter with that name.
+
+        Parameters
+        ----------
+        age_or_key : int or str
+            Age or key of parameter(s)
+
+
+        Returns
+        -------
+        Parameters or value
+            Parameters object with parameters that apply to age `age_or_key` or value of
+            parameter with name `age_or_key`.
+        """
+        if isinstance(age_or_key, int):
+            # return parameters at age
+            assert age_or_key < self._term_age
```

Comment:
It tends to be bad for to leave `asserts` in production code.
It would be better to throw an exception.